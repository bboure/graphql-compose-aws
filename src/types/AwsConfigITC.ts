import { schemaComposer } from 'graphql-compose';

export default schemaComposer.createInputTC({
  name: 'AwsConfig',
  fields: {
    accessKeyId: 'String',
    secretAccessKey: 'String',
    region: 'String',
  },
});
