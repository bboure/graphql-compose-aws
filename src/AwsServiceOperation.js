/* @flow */

import { upperFirst, GraphQLJSON } from 'graphql-compose';
import type {
  GraphQLFieldConfig,
  GraphQLFieldConfigArgumentMap,
  GraphQLFieldResolver,
  GraphQLOutputType,
} from 'graphql-compose/lib/graphql';
import { AwsParam, type ParamStructure } from './AwsParam';
import type { AwsShapes } from './AwsShapes';
import type { AwsSDK } from './AwsApiParser';
import AwsConfigITC from './types/AwsConfigITC';

export type ServiceOperationConfig = {
  http?: {
    method: string,
    requestUri: string,
  },
  input?: ParamStructure,
  output?: ParamStructure,
  alias?: string,
};

type OperationOpts<TContext> = {|
  serviceId: string,
  name: string,
  prefix: string,
  config: ServiceOperationConfig,
  shapes: AwsShapes<TContext>,
  awsSDK: AwsSDK,
|};

export class AwsServiceOperation<TContext> {
  prefix: string;
  name: string;
  serviceId: string;
  awsSDK: AwsSDK;
  args: GraphQLFieldConfigArgumentMap;
  config: ServiceOperationConfig;
  shapes: AwsShapes<TContext>;

  constructor(opts: OperationOpts<TContext>) {
    this.prefix = opts.prefix;
    this.name = opts.name;
    this.serviceId = opts.serviceId;
    this.awsSDK = opts.awsSDK;
    this.config = opts.config;
    this.shapes = opts.shapes;
  }

  getTypeName(): string {
    return `${this.prefix}${upperFirst(this.serviceId)}${upperFirst(this.name)}`;
  }

  getType(): GraphQLOutputType {
    if (this.config.output && this.config.output.type === 'structure') {
      const tc = AwsParam.convertOutputStructure(
        this.config.output,
        `${this.getTypeName()}Output`,
        this.shapes
      );

      // If bugous structure without fields, then return JSON
      if (Object.keys(tc.getFields()).length === 0) {
        return GraphQLJSON;
      }

      return tc.getType();
    }

    return GraphQLJSON;
  }

  getArgs(): GraphQLFieldConfigArgumentMap {
    if (!this.args) {
      if (!this.config.input) {
        this.args = {};
      } else {
        const itc = AwsParam.convertInputStructure(
          this.config.input,
          this.getTypeName(),
          this.shapes
        );

        let type;
        // If bugous structure without fields, then return JSON
        if (Object.keys(itc.getFields()).length === 0) {
          type = GraphQLJSON;
        } else {
          const hasRequiredFields = itc.getFieldNames().some(f => itc.isRequired(f));
          type = hasRequiredFields ? (itc.getTypeNonNull(): any) : itc.getType();
        }

        this.args = {
          input: {
            type,
          },
        };
      }

      this.args.config = {
        type: AwsConfigITC.getType(),
      };
    }

    return this.args;
  }

  getResolve(): GraphQLFieldResolver<any, any> {
    return ((source: any, args: { [argument: string]: any }) => {
      return new Promise((resolve, reject) => {
        const awsConfig = {
          ...(source && source.awsConfig),
          ...(args && args.config),
        };
        const service = new this.awsSDK[this.serviceId](awsConfig);
        service[this.name](args.input, (err, data) => {
          if (err) {
            reject(err);
          }
          resolve(data);
        });
      });
    }: any);
  }

  getFieldConfig(): GraphQLFieldConfig<any, any> {
    return {
      type: this.getType(),
      args: this.getArgs(),
      resolve: this.getResolve(),
    };
  }
}
