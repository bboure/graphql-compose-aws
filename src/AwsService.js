/* @flow */

import { schemaComposer, ObjectTypeComposer, upperFirst } from 'graphql-compose';
import type { GraphQLObjectType, GraphQLFieldConfig } from 'graphql-compose/lib/graphql';
import { AwsServiceMetadata, type ServiceMetadataConfig } from './AwsServiceMetadata';
import { AwsServiceOperation, type ServiceOperationConfig } from './AwsServiceOperation';
import { AwsShapes, type ShapesMap } from './AwsShapes';
import type { AwsSDK } from './AwsApiParser';
import AwsConfigITC from './types/AwsConfigITC';

export type ServiceConfig = {
  version: string,
  metadata: ServiceMetadataConfig,
  operations: {
    [name: string]: ServiceOperationConfig,
  },
  shapes: ShapesMap,
  paginators: any,
  waiters: any,
};

type ServiceOpts = {|
  serviceId: string,
  prefix: string,
  config: ServiceConfig,
  awsSDK: AwsSDK,
|};

export class AwsService<TContext> {
  prefix: string;
  serviceId: string;
  awsSDK: AwsSDK;
  tc: ?ObjectTypeComposer<any, TContext>;
  config: ServiceConfig;
  metadata: AwsServiceMetadata<TContext>;
  shapes: AwsShapes<TContext>;

  constructor(opts: ServiceOpts) {
    this.prefix = opts.prefix;
    this.serviceId = opts.serviceId;
    this.awsSDK = opts.awsSDK;
    this.config = opts.config;

    this.metadata = new AwsServiceMetadata(this.config.metadata);
    this.shapes = new AwsShapes(this.config.shapes, this.getTypeName());
  }

  getTypeName(): string {
    return `${this.prefix}${upperFirst(this.serviceId)}`;
  }

  static lowerFirst(name: string) {
    return name.charAt(0).toLowerCase() + name.slice(1);
  }

  getOperationNames(): string[] {
    return Object.keys(this.config.operations);
  }

  getOperation(name: string): AwsServiceOperation<TContext> {
    const operConfig = this.config.operations[name];
    if (!operConfig) {
      throw new Error(`Operation with name ${name} does not exist.`);
    }
    return new AwsServiceOperation({
      serviceId: this.serviceId,
      name: this.constructor.lowerFirst(name),
      prefix: this.prefix,
      config: operConfig,
      shapes: this.shapes,
      awsSDK: this.awsSDK,
    });
  }

  getTypeComposer(): ObjectTypeComposer<any, TContext> {
    if (!this.tc) {
      const fields = this.getOperationNames().reduce((res, name) => {
        res[this.constructor.lowerFirst(name)] = this.getOperation(name).getFieldConfig();
        return res;
      }, {});

      this.tc = schemaComposer.createObjectTC({
        name: this.getTypeName(),
        fields,
        description: this.metadata.getDescription(),
      });
    }
    return this.tc;
  }

  getType(): GraphQLObjectType {
    return this.getTypeComposer().getType();
  }

  getFieldConfig(): GraphQLFieldConfig<any, any> {
    return {
      type: this.getType(),
      args: {
        config: {
          type: AwsConfigITC.getType(),
          description: 'Will override default configs for aws-sdk.',
        },
      },
      resolve: (source, args) => ({
        awsConfig: { ...(source && source.awsConfig), ...(args && args.config) },
      }),
    };
  }
}
