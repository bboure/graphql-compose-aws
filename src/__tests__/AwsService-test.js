/* @flow */

import { AwsService } from '../AwsService';
import { AwsServiceOperation } from '../AwsServiceOperation';
import s3Cfg from '../__mocks__/s3-2006-03-01.json';
import AwsConfigITC from '../types/AwsConfigITC';

describe('AwsService', () => {
  const s3 = new AwsService({
    serviceId: 'S3',
    prefix: 'Aws',
    config: s3Cfg,
    awsSDK: ({}: any),
  });

  it('getTypeName()', () => {
    expect(s3.getTypeName()).toBe('AwsS3');
  });

  it('getOperationNames()', () => {
    expect(s3.getOperationNames()).toMatchSnapshot();
  });

  it('getOperation()', () => {
    const oper = s3.getOperation('CreateBucket');
    expect(oper).toBeInstanceOf(AwsServiceOperation);
    expect(oper.getTypeName()).toBe('AwsS3CreateBucket');
    expect(oper.getArgs().input).toBeDefined();
  });

  it('getTypeComposer()', () => {
    const tc = s3.getTypeComposer();
    expect(tc.getTypeName()).toBe('AwsS3');
    expect(tc.getFieldNames()).toContain('createBucket');
    expect(tc.getFieldNames()).toContain('deleteObject');
  });

  it('getType()', () => {
    const type = s3.getType();
    expect(type.name).toBe('AwsS3');
  });

  it('getFieldConfig()', () => {
    const fc: any = s3.getFieldConfig();
    expect(fc.type).toBe(s3.getType());
    expect(fc.args.config.type).toBe(AwsConfigITC.getType());
    expect(fc.resolve()).toEqual({ awsConfig: {} });
  });
});
