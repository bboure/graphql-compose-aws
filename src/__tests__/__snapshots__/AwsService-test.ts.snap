// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AwsService getOperationNames() 1`] = `
Array [
  "AbortMultipartUpload",
  "CompleteMultipartUpload",
  "CopyObject",
  "CreateBucket",
  "CreateMultipartUpload",
  "DeleteBucket",
  "DeleteBucketAnalyticsConfiguration",
  "DeleteBucketCors",
  "DeleteBucketEncryption",
  "DeleteBucketInventoryConfiguration",
  "DeleteBucketLifecycle",
  "DeleteBucketMetricsConfiguration",
  "DeleteBucketPolicy",
  "DeleteBucketReplication",
  "DeleteBucketTagging",
  "DeleteBucketWebsite",
  "DeleteObject",
  "DeleteObjectTagging",
  "DeleteObjects",
  "GetBucketAccelerateConfiguration",
  "GetBucketAcl",
  "GetBucketAnalyticsConfiguration",
  "GetBucketCors",
  "GetBucketEncryption",
  "GetBucketInventoryConfiguration",
  "GetBucketLifecycle",
  "GetBucketLifecycleConfiguration",
  "GetBucketLocation",
  "GetBucketLogging",
  "GetBucketMetricsConfiguration",
  "GetBucketNotification",
  "GetBucketNotificationConfiguration",
  "GetBucketPolicy",
  "GetBucketReplication",
  "GetBucketRequestPayment",
  "GetBucketTagging",
  "GetBucketVersioning",
  "GetBucketWebsite",
  "GetObject",
  "GetObjectAcl",
  "GetObjectTagging",
  "GetObjectTorrent",
  "HeadBucket",
  "HeadObject",
  "ListBucketAnalyticsConfigurations",
  "ListBucketInventoryConfigurations",
  "ListBucketMetricsConfigurations",
  "ListBuckets",
  "ListMultipartUploads",
  "ListObjectVersions",
  "ListObjects",
  "ListObjectsV2",
  "ListParts",
  "PutBucketAccelerateConfiguration",
  "PutBucketAcl",
  "PutBucketAnalyticsConfiguration",
  "PutBucketCors",
  "PutBucketEncryption",
  "PutBucketInventoryConfiguration",
  "PutBucketLifecycle",
  "PutBucketLifecycleConfiguration",
  "PutBucketLogging",
  "PutBucketMetricsConfiguration",
  "PutBucketNotification",
  "PutBucketNotificationConfiguration",
  "PutBucketPolicy",
  "PutBucketReplication",
  "PutBucketRequestPayment",
  "PutBucketTagging",
  "PutBucketVersioning",
  "PutBucketWebsite",
  "PutObject",
  "PutObjectAcl",
  "PutObjectTagging",
  "RestoreObject",
  "UploadPart",
  "UploadPartCopy",
]
`;
